### PR TITLE
Fix engineFreeModel crashes

### LINKED ISSUE #3191: engineFreeModel Crashes client if object destroyed
### Describe the bug

The client just crashes when trying to call engineFreeModel if allocated object was be destroyed

### Steps to reproduce

1. Request model with engineRequestModel
2. Load and import any txd/dff with requested model
3. Сreate an object with imported model near you
4. Destroy the object
5. use engineFreeModel

### Version

v1.6-release-22215

### Additional context

[crash_test.zip](https://github.com/multitheftauto/mtasa-blue/files/12691238/crash_test.zip)

Test resource. Just use command `testCrash`

### Relevant log output

_No response_

### Security Policy

- [X] I have read and understood the Security Policy and this issue is not security related.

### PR DESCRIPTION
**Pull request by TheNormalnij**
Fixes #3191

### CODE REVIEW

File: Client/mods/deathmatch/logic/CClientModel.cpp [hunk 1]
```diff
@@ -78,30 +78,49 @@ bool CClientModel::Allocate(ushort usParentID)
     return false;
 }
 
+// You can call it only in destructor for DFF.
```
Comment (botder) [inline Client/mods/deathmatch/logic/CClientModel.cpp:81]: Shouldn't you mark this method private in such case?

File: Client/mods/deathmatch/logic/CClientModel.cpp [hunk 2]
```diff
@@ -78,30 +78,49 @@ bool CClientModel::Allocate(ushort usParentID)
...
 bool CClientModel::Deallocate()
 {
     if (!m_bAllocatedByUs)
```
Comment (botder) [inline Client/mods/deathmatch/logic/CClientModel.cpp:84]: Looking over the class...wouldn't it make more sense if `m_bAllocatedByUs = true;` (line 28) was done only in the case where we actually allocated a new model? You won't need that if-block around `pModelInfo->DeallocateModel();`.

File: Client/mods/deathmatch/logic/CClientModel.cpp [hunk 3]
```diff
@@ -78,30 +78,49 @@ bool CClientModel::Allocate(ushort usParentID)
...
+        // Remove model info
+        pModelInfo->DeallocateModel();
+    }
+
+    return true;
+}
+
+void CClientModel::RestoreEntitiesUsingThisModel()
+{
+    CModelInfo* pModelInfo = g_pGame->GetModelInfo(m_iModelID, true);
```
Comment (botder) [inline Client/mods/deathmatch/logic/CClientModel.cpp:104]: Don't you also need an `if (!m_bAllocatedByUs)` here?

Comment (TheNormalnij) [reply Client/mods/deathmatch/logic/CClientModel.cpp:104]: ~~I think we can remove m_bAllocatedByUs from this class completely~~

File: Client/mods/deathmatch/logic/CClientModel.cpp (hunk 1 above)
Comment (TheNormalnij) [reply Client/mods/deathmatch/logic/CClientModel.cpp:81]: It is still possible to use it with TXD

File: Client/mods/deathmatch/logic/CClientModel.cpp (hunk 2 above)
Comment (TheNormalnij) [reply Client/mods/deathmatch/logic/CClientModel.cpp:84]: Some classes CClientPed/CClientVehicle/CClientObject have field `std::shared_ptr<CClientModel> m_clientModel`. I think it is used to make deallocation after unloading a model from stream.

File: Client/mods/deathmatch/logic/CClientModel.cpp (hunk 3 above)
Comment (TheNormalnij) [reply Client/mods/deathmatch/logic/CClientModel.cpp:104]: Here can be check. But this is unnecessary for now. This code used only for deallocation.

**Review by botder — APPROVED** [review]

**Pull request merged**